src/main/scala/epic/logo/Trainer.scala

```diff
@@ -157,32 +162,33 @@ object Trainer {
       val trainer = newL1MaxMarginTrainer(new MulticlassOneSlackOracleInferencer(oracler),
         new MulticlassOneSlackLossAugmentedArgmaxInferencer(argmaxer, initialConstraint),
         oneSlackIterCallBack, C, maxNumIters, opts)
-      trainer.train(Seq(new Instance(data)),
-        new Weights(space.zeroLike(initialConstraint)))
+      trainer.train(Seq(data), new Weights(space.zeroLike(initialConstraint)))
     } else {
       val trainer = newL1MaxMarginTrainer(oracler, argmaxer, iterationCallback, C, maxNumIters, opts)
-      trainer.train(data.map(new Instance[LabeledDatum[L, F], W](_)),
-        new Weights(space.zeroLike(initialConstraint)))
+      trainer.train(data, new Weights(space.zeroLike(initialConstraint)))
     }
     new MulticlassClassifier[L, F, W](weights, argmaxer)
   }
 }
 
-case class Trainer[T, W](convergenceChecker: ConvergenceChecker[W],
-                         iterationCallback: IterationCallback[T, W], val decoder: Decoder[T, W],
-                         updater: Updater[W],
-                         opts: LogoOpts = new LogoOpts(),
-                         initialConstraintAndAlpha: Option[((W, Double), Double)] = None,
-                         online: Boolean = false,
-                         average: Boolean = false)(implicit space: MutableInnerProductModule[W, Double])
+case class Trainer[T, W, OracleS, MaxerS](
+  convergenceChecker: ConvergenceChecker[W],
+  iterationCallback: IterationCallback[T, W, OracleS, MaxerS],
+  decoder: Decoder[T, W, OracleS, MaxerS],
+  updater: Updater[W],
+  opts: LogoOpts = new LogoOpts(),
+  initialConstraintAndAlpha: Option[((W, Double), Double)] = None,
+  online: Boolean = false,
+  average: Boolean = false)(implicit space: MutableInnerProductModule[W, Double])
     extends SerializableLogging {
   import space._
 
   final val eps = opts.constraintEpsilon
   final val numInnerOptimizationLoops = if (online) 1 else opts.numInnerOptimizationLoops
   final val numOuterOptimizationLoops = if (online) 0 else opts.numOuterOptimizationLoops
 
-  def train(data : Seq[Instance[T, W]], initWeights : Weights[W]) = {
+  def train(rawData : Seq[T], initWeights : Weights[W]) = {
+    val data = rawData.map(datum => Instance[T, W](datum))
```

Comment:
ah right, i forgot.  I don't think it's a good name, but whatever, i don't know what a better name is...